subsys/fs/nvs/nvs.c

```diff
@@ -0,0 +1,804 @@
+	addr = (fs->write_location & ~(fs->sector_size - 1));
+	_nvs_addr_advance(fs, &addr, fs->sector_size);
+
+	/* initialize the new sector */
+	rc = _nvs_sector_init(fs, addr);
+	if (rc) {
+		goto out;
+	}
+
+	/* data copy */
+	addr = (fs->write_location & ~(fs->sector_size - 1));
+	_nvs_addr_advance(fs, &addr, fs->sector_size);
+	/* Do we need to advance the entry_sector, if so we need to copy */
+	if ((addr & ~(fs->sector_size - 1)) ==
+		fs->entry_sector * fs->sector_size) {
+		SYS_LOG_DBG("Starting data copy...");
+		_nvs_entry_sector_advance(fs);
+		rc = _nvs_gc(fs, addr);
+		if (rc) {
+			SYS_LOG_DBG("Quit data copy - gc error");
+			goto out;
+		}
+		if (_nvs_flash_erase(fs, addr, fs->sector_size)) {
+			rc = NVS_ERR_FLASH;
+			SYS_LOG_DBG("Quit data copy - flash erase error");
+			goto out;
+		}
+		SYS_LOG_DBG("Done data copy - no error");
+	}
+
+	rc = NVS_OK;
+
+out:
+	k_mutex_unlock(&fs->nvs_lock);
+	return rc;
+}
+
+int nvs_clear(struct nvs_fs *fs)
+{
+	int rc;
+	off_t addr;
+
+	k_mutex_lock(&fs->nvs_lock, K_FOREVER);
+
+	for (addr = 0; addr < fs->sector_count * fs->sector_size;
+		addr += fs->sector_size) {
+		rc = _nvs_flash_erase(fs, addr, fs->sector_size);
+		if (rc) {
+			goto out;
+		}
+	}
+	rc = NVS_OK;
+
+out:
+	k_mutex_unlock(&fs->nvs_lock);
+	return rc;
+}
+
+int nvs_flash_read(struct nvs_fs *fs, off_t offset, void *data, size_t len)
+{
+
+	if (_nvs_bd_check(fs, offset, len)) {
+		return NVS_ERR_ARGS;
+	}
+
+	if (flash_read(fs->flash_device, fs->offset + offset, data, len)) {
+		/* flash read error */
+		return NVS_ERR_FLASH;
+	}
+	return NVS_OK;
+}
+
+int nvs_flash_write(struct nvs_fs *fs, off_t offset, const void *data,
+	size_t len)
+{
+
+	if (_nvs_bd_check(fs, offset, len)) {
+		return NVS_ERR_ARGS;
+	}
+
+	if (flash_write_protection_set(fs->flash_device, 0)) {
+		/* flash protection set error */
+		return NVS_ERR_FLASH;
+	}
+	if (flash_write(fs->flash_device, fs->offset + offset, data, len)) {
+		/* flash write error */
+		return NVS_ERR_FLASH;
+	}
+	(void) flash_write_protection_set(fs->flash_device, 1);
+	/* don't mind about this error */
+	return NVS_OK;
+}
+
+
+int nvs_write(struct nvs_fs *fs, struct nvs_entry *entry, const void *data)
+{
+	int rc;
+	struct nvs_entry stored_entry;
+
+	if ((entry->id == NVS_ID_EMPTY) || (entry->id == NVS_ID_SECTOR_END)) {
+		return NVS_ERR_ARGS;
+	}
+	if (entry->len > fs->max_len) {
+		return NVS_ERR_LEN;
+	}
+	if ((entry->len > 0) && (data == NULL)) {
+		return NVS_ERR_ARGS;
+	}
+	/* check we are not trying to write the same data */
+	stored_entry.id = entry->id;
+	rc = nvs_get_last_entry(fs, &stored_entry);
+	if (rc == NVS_OK) {
+		if (stored_entry.len == _nvs_len_in_flash(fs, entry->len)) {
+			if (memcmp(data,
+				(void *) (stored_entry.data_addr + fs->offset),
+				entry->len) == 0) {
+				return NVS_OK;
+			}
+		}
+	}
+	/* new data or data has changed */
+	while (1) {
+		rc = nvs_append(fs, entry);
+		if (rc) {
+			if (rc == NVS_ERR_NOSPACE) {
+				if (nvs_rotate(fs)) {
```

Comment:
@Laczen Assume that NVS I full - so nvs_appends() will always failing, then this routine will keeps calling nvs_rotate() which cause loping here. Am I right?

Reply:
@nvlsianpu You are right, I think the best way to avoid this is to keep track of how much of the file system is used and return NVS_ERR_NOSPACE if there is insufficient space (before calling nvs_rotate). This would give the extra of being able to report the space usage.

I could also limit the amount of calls to nvs_rotate() but I don't think this is a good idea since calling a nvs_write again would restart this proces and so this could get into a loop again.

Reply:
What you proposed sounds good to me.
Second option is to count rotations and once storage is full use a storage_full flag to discard write right after was called.

Reply:
@Laczen Hello,
I believe that if you keep track of how much space is used and determine that an entry can fit you can call nvs_rotate() until nvs_append() succeeds.

Reply:
@nvlsianpu Your idea is good too, and a lot simpler. If there are more rotations required then the number of sectors, storage is full. A call of nvs_delete() could reset the storage_full flag, at least if the nvs_delete() call is succesfull.

@lemrey That was the idea, however I do like the simplicity of the proposal by nvlsianpu.

Reply:
@Laczen Great, you understand exactly about I was thinking. @lemrey Anyway - NVS storage consumption feature might be added in any subsequent patch - on it will be required ba a user case.

Reply:
@Laczen One thing that comes to mind is how to deal with the situation where the filesystem is full and one attemps to delete an entry, since deleting an entry writes to a new, unused location in flash.

Reply:
@Laczen There might be a few options, the one I am thinking about would be to introduce a new field in the entry header to keep track of whether the entry is valid or dirty (deleted). This would also simplify gc quite a bit, since it would only need to read the entries in the oldest sector, as opposed to reading every other entry in every other sector for each entry in the oldest sector (!).

Reply:
@lemrey You are right, this is why I added that the nvs_delete() call should be succesfull. There could be a case where a delete is not possible because there is not even space to add the item header and footer for a delete, in this case the file system will remain full but return a delete error. You should be really unlucky to get in such a case as there is always unused space at the end of a sector. Anyhow, you should never use NVS in such a way that the filesystem is full (this means all different items taking up all but 1 sectors) this will destroy flash to fast. You could get there by accident and in this case it should not destroy the flash.

@nvlsianpu, @lemrey I am thinking about making it a little bit more complex:

Instead of defining a storage_full I would define a space_available variable: space_available is initialized equal to max_len. If storage is full (through rotation count) space_available is set to zero. As long as space_available is smaller then max_len no writes are allowed. nvs_delete() calls increase space_available by the deleted size up to max_len. As soon as space_available reaches max_len writes will be allowed again.

Reply:
@lemrey, I have considered using a field like this in the design (there is free space in the footer), this would require rewriting in flash and might not be supported in all flash devices so I decided not to do this.